cdc/cdc.cc

```diff
@@ -194,4 +195,194 @@ future<> setup(service::storage_proxy& proxy, schema_ptr s) {
+    std::reference_wrapper<const locator::abstract_replication_strategy> _replication_strategy;
+    // The assumption here is that all nodes in the cluster are using the
+    // same murmur3_partitioner_ignore_msb_bits. If it's not true, we will
+    // have to gossip it around together with shard_count.
+    unsigned _ignore_msb_bits;
+    streams_type _streams;
+
+    clustering_key set_pk_columns(const partition_key& pk, int batch_no, mutation& m) const {
+        const auto log_ck = clustering_key::from_exploded(
+                *m.schema(), { _decomposed_time, int32_type->decompose(batch_no) });
+        auto pk_value = pk.explode(*_schema);
+        size_t pos = 0;
+        for (const auto& column : _schema->partition_key_columns()) {
+            assert (pos < pk_value.size());
+            auto cdef = m.schema()->get_column_definition(to_bytes("_" + column.name()));
+            auto value = atomic_cell::make_live(*column.type,
+                                                _time.timestamp(),
+                                                bytes_view(pk_value[pos]));
+            m.set_cell(log_ck, *cdef, std::move(value));
+            ++pos;
+        }
+        return log_ck;
+    }
+    partition_key stream_id(const schema& s, const net::inet_address& ip, unsigned int shard_id) const {
+        auto it = _streams.find(std::make_pair(ip, shard_id));
+        if (it == std::end(_streams)) {
+                throw std::runtime_error(format("No stream found for node {} and shard {}", ip, shard_id));
+        }
+        return partition_key::from_exploded(s, { uuid_type->decompose(it->second) });
+    }
+public:
+    transformer(database& db, schema_ptr s, streams_type streams)
+        : _schema(std::move(s))
+        , _log_schema(db.find_schema(_schema->ks_name(), log_name(_schema->cf_name())))
+        , _time(utils::UUID_gen::get_time_UUID())
+        , _decomposed_time(timeuuid_type->decompose(_time))
+        , _token_metadata(service::get_local_storage_service().get_token_metadata())
+        , _replication_strategy(db.find_keyspace(_schema->ks_name()).get_replication_strategy())
+        , _ignore_msb_bits(db.get_config().murmur3_partitioner_ignore_msb_bits())
+        , _streams(std::move(streams))
+    { }
+
+    mutation transform(const mutation& m) const {
+        auto& t = m.token();
+        auto&& eps = _replication_strategy.get().calculate_natural_endpoints(t, _token_metadata.get());
+        if (eps.empty()) {
+            throw std::runtime_error(format("No primary replica found for key {}", m.decorated_key()));
+        }
+        auto shard_id = dht::murmur3_partitioner::shard_of(
+                t, locator::i_endpoint_snitch::get_local_snitch_ptr()->get_shard_count(eps[0]), _ignore_msb_bits);
+        mutation res(_log_schema, stream_id(*_log_schema, eps[0].addr(), shard_id));
+        set_pk_columns(m.key(), 0, res);
+        return res;
+    }
+};
+
+class streams_builder {
+    schema_ptr _schema;
```

Comment:
suggestion: this can be a `const schema&`, since `streams_builder` doesn't have to manage the `schema`'s lifetime: it's only a temporary object that doesn't live across different tasks.

The same applies to `transformer`, but changing that would require a bit more effort.